src/encoding.ts

```diff
@@ -0,0 +1,379 @@
+				if (lowSurrogate >= 0xDC00 && lowSurrogate < 0xDFFF) {
+					encodedChar = 0x010000 + (encodedChar - 0xD800) * 0x0400 + (lowSurrogate - 0xDC00);
+				}
+				else {
+					i--;
+				}
+			}
+
+			if (encodedChar < 0x80) {
+				buffer[position++] = encodedChar;
+			}
+			else {
+				if (encodedChar < 0x800) {
+					buffer[position++] = ((encodedChar >> 0x06) & 0x1F) | 0xC0;
+				}
+				else if (encodedChar < 0x010000) {
+					if (encodedChar >= 0xD800 && encodedChar <= 0xDFFF) {
+						throw Error('Surrogate is not a scalar value');
+					}
+
+					buffer[position++] = ((encodedChar >> 0x0C) & 0x0F) | 0xE0;
+					buffer[position++] = ((encodedChar >> 0x06) & 0x3F) | 0x80;
+				}
+				else if (encodedChar < 0x200000) {
+					buffer[position++] = ((encodedChar >> 0x12) & 0x07) | 0xF0;
+					buffer[position++] = ((encodedChar >> 0x0C) & 0x3F) | 0x80;
+					buffer[position++] = ((encodedChar >> 0x06) & 0x3F) | 0x80;
+				}
+				buffer[position++] = (encodedChar & 0x3F) | 0x80;
+			}
+		}
+
+		buffer.length = position;
+
+		return buffer;
+	}
+
+	/**
+	 * Decode a Utf8 encoded byte buffer into a string
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	static decode(data: ByteBuffer): string {
+		if (!validateDecodingArgs(data)) {
+			return '';
+		}
+
+		let decoded = '';
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let byte1 = data[i] & 0xFF;
+
+			if ((byte1 & 0x80) === 0) {
+				decoded += decodeUtf8EncodedCodePoint(byte1);
+			}
+			else if ((byte1 & 0xE0) === 0xC0) {
+				let byte2 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte2);
+				byte2 = byte2 & 0x3F;
+				let encodedByte = ((byte1 & 0x1F) << 0x06) | byte2;
+				decoded += decodeUtf8EncodedCodePoint(encodedByte, [0x80, Infinity]);
+			}
+			else if ((byte1 & 0xF0) === 0xE0) {
+				let byte2 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte2);
+				byte2 = byte2 & 0x3F;
+
+				let byte3 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte3);
+				byte3 = byte3 & 0x3F;
+
+				let encodedByte = ((byte1 & 0x1F) << 0x0C) | (byte2 << 0x06) | byte3;
+				decoded += decodeUtf8EncodedCodePoint(encodedByte, [0x0800, Infinity], true);
+			}
+			else if ((byte1 & 0xF8) === 0xF0) {
+				let byte2 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte2);
+				byte2 = byte2 & 0x3F;
+
+				let byte3 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte3);
+				byte3 = byte3 & 0x3F;
+
+				let byte4 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte4);
+				byte4 = byte4 & 0x3F;
+
+				let encodedByte = ((byte1 & 0x1F) << 0x0C) | (byte2 << 0x0C) | (byte3 << 0x06) | byte4;
+				decoded += decodeUtf8EncodedCodePoint(encodedByte, [0x010000, 0x10FFFF]);
+			}
+			else {
+				validateUtf8EncodedCodePoint(byte1);
+			}
+		}
+
+		return decoded;
+	}
+}
+
+/**
+ * Hex codec that provides facilities for encoding a string (or string coerced object)
+ * into an Hex encoded byte buffer as well as decoding an Hex encoded byte buffer
```

Comment:
The original goal was to convert a hex encoded string, like "DEADBEEF01", into a byte buffer, and vice versa, rather than to store a string in a particular binary format.